[windows] implement regular windows and supporting method channels

mattkae opened the pull request:
This pull request only makes changes to the windows embedder API and wires a flag through the framework. This does not include any user-facing API changes.

Design doc: https://docs.google.com/document/d/1eQG-IS7r4_S9_h50MY_hSGwUVtgSTRSLzu7MLzPNd2Y/edit?tab=t.0

## To test
1. Check out [this branch in canonical/flutter](https://github.com/canonical/flutter/pull/36) (this includes changes to the framework which are not included here on purpose for reviewer's sake!)
2. Build the engine
3. Then run the multi window reference application:
   ```
   cd examples/multi_window_ref_app/
   flutter run --debug --local-engine-src-path <PATH_TO_ENGINE_SRC> --local-engine host_debug_unopt --local-engine-host host_debug_unopt .\lib\main.dart --enable-multi-window
   ```
Don't forget the `--enable-multi-window` flag or it will not work

## What's new?

### Framework
- Added a new `--enable-multi-window` flag to turn on multi window features

### Engine
- Implemented regular windows in the windows embedder

![image](https://github.com/user-attachments/assets/dcfe0fd9-d631-4887-a92f-c6202e9a4fea)


## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

mattkae committed:
feature: implement createRegularWindow and add it to the reference app

justinmc commented:
@mattkae Should we put all the subsequent PRs in draft mode for now (assuming they are not yet ready for review)? Just so other triagers don't get confused.

Thanks for all of this work by the way!

CC @loic-sharma @cbracken Any idea who should take on getting all of this reviewed?

mattkae commented:
> @mattkae Should we put all the subsequent PRs in draft mode for now (assuming they are not yet ready for review)? Just so other triagers don't get confused.

If you would prefer that, I can put the others in draft mode to avoid confusion :+1: 
 
> Thanks for all of this work by the way!

No problem :) Looking forward to working towards landing it !

loic-sharma commented:
> CC @loic-sharma @cbracken Any idea who should take on getting all of this reviewed?

I suspect @goderbauer would be the best person to review once he's back :)

mattkae committed:
wip: refactor of the foundational code

mattkae committed:
wip: refactored the mutlti window reference app to use new API

mattkae committed:
wip: able to create windows in the reference application again

mattkae committed:
Selection and rerenders are functioning appropriately

mattkae committed:
Able to remove an application, albeit with some bugs + refactored how settings are donw

mattkae committed:
Able to add and remove windows

mattkae committed:
All windows are now showing in the list

mattkae committed:
Added the ability to listen for errors and respond to them accordingly

mattkae committed:
RegularWindowMetadata and WindowMetadata are useless

mattkae committed:
The future should be set in setState

mattkae committed:
Properly handling the event where we need to destroy the widget during a dispose

mattkae committed:
Fix issue where setState returned a future

mattkae committed:
setState is ssynchronous, whoops!

mattkae committed:
Refactored window.dart to be more concise for the initial PR

mattkae committed:
Fixed issue where we might accidentally dispose of a window twice

mattkae committed:
Making the WindowMetadata public and only refresing the relevant fields when settings change

mattkae committed:
It does not make sense for the modify method to be an override since different window types will have different values

mattkae committed:
bugfix: able to select the initial wndow in the table

mattkae committed:
Reimplement the rotating cube in the demo app

mattkae committed:
PR feedback

mattkae committed:
Rendering correct text

mattkae committed:
Reverting back to the old runner code

mattkae committed:
Update runner code to work with the latest & greatest

hbatagelo committed:
Add command-line option for multi-window support and refactor runner code

mattkae commented:
@goderbauer @loic-sharma If you are interested, this PR now reflect "declarative API" that we settled on in the spec. The tests still need to be updated, but this should give you a sense of the general look-and-feel of the API as we've described it so far in the document.

mattkae committed:
feature: adding full client-side tests for regular window creation  (#14)

feature: adding full client-side tests for regular window creation

hbatagelo committed:
Add foundational support for Windows multi-window in the engine (#15)

bkonyi commented:
@goderbauer @jtmcdole can you find the right reviewers for this non-tool portions of this PR?

hbatagelo committed:
Refactor method channel for regular windows

hbatagelo committed:
Review feedback + align code style with guidelines

hbatagelo committed:
Add FML as a dependency to common_cpp_core; fix incorrect include

hbatagelo committed:
Improve comments and code style consistency

mattkae committed:
Merge pull request #19 from canonical/refactor-method-channel

Refactor method channel for regular windows

mattkae committed:
Refactor regular windows (#20)

* refactor: the RegularWindowController now holds all initial values and is solely responsible for creating and destroying a window

* Updated multi window reference app and window tests

mattkae committed:
refactor: remove references to WindowingApp from the window.dart API (#21)

* refactor: remove references to WindowingApp from the window.dart API

* bugfix: notify listeners when values change on a window controller

mattkae committed:
refactor: raw create/destroy methods are now hidden + renamed WindowContext to WindowControllerContext and made it provide the controller (#22)

mattkae committed:
feature: implemented modifyRegular + fixed a bug where a widget tree was mandatory on accident

mattkae committed:
feature: tidy up all of the documentation around regular windows + simplify parts of the API (#25)

mattkae committed:
refactor: distinguish a WindowCreationResult from a particular RegularWindowCreationResult

mattkae committed:
minor: fix spelling mistake

mattkae committed:
git is confused and not recognizing my pushes...?

hbatagelo committed:
Add support for modifyRegular in the embedder and update examples/tests

hbatagelo committed:
Better ModifyRegularWindow test

robert-ancell commented on engine/src/flutter/shell/platform/windows/windowing_handler.cc:
```diff
@@ -0,0 +1,267 @@
+constexpr char kViewIdKey[] = "viewId";
+
+// Error codes used for responses.
+constexpr char kInvalidValueError[] = "Invalid Value";
```
In other channels this would be:
```
constexpr char kBadArgumentsError[] = "Bad Arguments";
```

hbatagelo replied:
Done. Updated to `kBadArgumentsError`

hbatagelo committed:
Add inequality operator for flutter::Size

hbatagelo committed:
Add SetState and remove unused GetState

hbatagelo committed:
Log error messages during window creation, window change, and destruction

hbatagelo committed:
Split up tests for modifying the window and add a test for changing the window state

mattkae committed:
Merge pull request #24 from canonical/regular-windows-modify

feature: implemented modifyRegular + fixed a bug where a widget tree was mandatory on accident

mattkae committed:
Merge pull request #28 from canonical/example/adding-edit

example: adding a modificatione example to the multi window reference app

mattkae committed:
Merge remote-tracking branch 'upstream/master' into foundation

mattkae committed:
example: adding a modificatione example to the multi window reference app

hbatagelo committed:
Use kBadArgumentsError

mattkae committed:
feedback: WindowControllerContext.of now returns the WindowController instead of the context

hbatagelo committed:
Add support for windows created in the runner

hbatagelo committed:
Use SetProp/GetProp for all windows + tweak comments for consistency

hbatagelo committed:
Fix checks and assertions; early forward to lifecycle manager

hbatagelo committed:
Ensure the host window controller is destroyed before the lifecycle manager

hbatagelo committed:
Merge pull request #32 from canonical/use-runner-windows

Add support for windows created in the runner

hbatagelo committed:
Merge remote-tracking branch 'upstream/master' into foundation

mattkae committed:
Merge branch 'master' into foundation

mattkae committed:
Merge branch 'master' into foundation

knopp commented:
I am wondering if `--enable-multi-window` is the best way to go about this. Is there a reason why this would be a flag that has to be specified for each build, rather then either `flutter config` or enabled simply through embedder API? The launcher needs to be modified for multi-window anyway.

mattkae committed:
Revert all API changes to the framework

knopp commented on engine/src/flutter/shell/platform/common/windowing.h:
```diff
@@ -0,0 +1,105 @@
+  std::optional<WindowState> state;
+};
+
+}  // namespace flutter
```
I don't know how I feel about this. Especially when looking at later PRs, which seem to put quite a bit of additional logic here, like [positioning for popovers](https://github.com/canonical/flutter/blob/popup-menu/engine/src/flutter/shell/platform/common/windowing.cc). Yes, we share some code between embedders already, but this seems very different as it implements a specific algorithm.

We have formal embedder API, "informal" platform channels, and now we have algorithms that embedders are expected to implement, placed at arbitrary location in the engine. What about out of tree embedders?

Wouldn't it be better if something like [this](https://github.com/canonical/flutter/blob/popup-menu/engine/src/flutter/shell/platform/common/windowing.cc) was implemented in dart instead?

robert-ancell replied:
In the case of positioning, modern compositors do this themselves, as an application never has all the context required to do this correctly (or is not allowed the context to maintain application isolation). For compositors that do not do this, the application needs to perform this, thus each compositor needs to implement this logic.

The difficulty in making shared dart code is the embedder is already running native code when it needs it. If we were to make this code run before requests reach the embedder it would make the API more complicated. In all cases, the shared code is more of a set of helper functions, as the behaviour always depends on the compositor in use.

knopp replied:
I don't quite understand how this addresses my concern. There is no precedent for sharing logic between embedders in-tree like this and I think this definitely needs further discussion.

I have proposal for synchronous ffi-based API https://docs.flutter.dev/go/multi-window-api-ffi. With synchronous API we can easily move all this to Dart code so that it can be accessed by all embedders, not just the "blessed" ones.

robert-ancell replied:
Is this not similar to the shared text input model code?

knopp replied:
Very vaguely. TextInputModel doesn't implement particular algorithm, at least not in a way where all embedders rely on. For example on macOS we're only using a tiny subset of it. iOS or Android is not using it at all. So it's more of a utility class rather than defining logic that embedders are supposed to follow.

Also text input is one of the most problematic parts of Flutter, and if we were starting from scratch we would be likely using a very different approach, see https://github.com/flutter/flutter/issues/150460. So modeling new code after text input doesn't seem like the best idea.

Saviq replied:
That's actually how this is meant to be used - as helpers - depending on platform, the embedder might use e.g. the positioner information itself (like on Windows) - on others (like Wayland) that context will be passed on verbatim to the compositor, so there's no point in making those calculations in Flutter at all. On others yet, child windows might not be supported altogether (mobile platforms, automotive).

The implementation here will only be shared between platforms that can meaningfully make use of it (floating window desktops - which sure, are the majority of them, but not all). Other platforms should default to not supporting child windows until/unless they grow support for at least a subset of them.

robert-ancell commented on engine/src/flutter/shell/platform/windows/flutter_host_window_controller.cc:
```diff
@@ -0,0 +1,229 @@
+
+// Names of the messages sent by the controller in response to window events.
+constexpr char kOnWindowChangedMethod[] = "onWindowChanged";
+constexpr char kOnWindowCreatedMethod[] = "onWindowCreated";
```
This seems to be unused.

robert-ancell commented on engine/src/flutter/shell/platform/windows/flutter_host_window_controller.cc:
```diff
@@ -0,0 +1,229 @@
+      map.insert(
+          {EncodableValue(kSizeKey),
+           EncodableValue(EncodableList{EncodableValue(size->width()),
+                                        EncodableValue(size->height())})});
```
How does this compare with `FlutterEngineSendWindowMetricsEvent`? Can the size be different to the size of the view (i.e. the Flutter widget is combined with other native widgets). Can the application make use of the size information?

hbatagelo replied:
They are essentially the same size. `onWindowChanged` sends the size of the client rectangle of the top-level window, which corresponds to the size of root view in logical coordinates.

robert-ancell reviewed (approved):
My Windows knowledge is pretty limited, but the code looks good to me. The only serious thing I noticed is the interaction with the window size and the view size, which might just need some clarification to ensure developers don't mix up the two.

robert-ancell commented on engine/src/flutter/shell/platform/windows/flutter_host_window_controller.h:
```diff
@@ -0,0 +1,101 @@
+  LRESULT HandleMessage(HWND hwnd, UINT message, WPARAM wparam, LPARAM lparam);
+
+  // Sets the method channel through which the controller will send the window
+  // events "onWindowCreated", "onWindowDestroyed", and "onWindowChanged".
```
onWindowCreated doesn't seem to exist anymore.

knopp commented:
What is the reason for trying to merge this without the appropriate framework API? This is tightly coupled to new framework API, I don't quite understand how this can be reviewed and merged separately before approving the framework part.

mattkae commented:
> What is the reason for trying to merge this without the appropriate framework API? This is tightly coupled to added framework API, I don't quite understand how this can be reviewed and merged separately before approving the framework part.

The reason is that the Google folks don't want to land a public API until (1) all multi-view bugs have been fixed and (2) we are certain that the API changes are more-or-less final.

@loic-sharma can provide more details, but we settled on this in the end to make things move faster :)

knopp commented on engine/src/flutter/shell/platform/windows/flutter_host_window_controller.cc:
```diff
@@ -0,0 +1,229 @@
+                                                   WPARAM wparam,
+                                                   LPARAM lparam) {
+  switch (message) {
+    case WM_NCDESTROY: {
```
Window destroying flow probably needs some thought? A very common scenario is displaying a prompt when closing the window (i.e. to prevent closing unsaved documents).

weltkante replied:
WM_NCDESTROY is just for cleanup of resources after closing, the window is already closed and being torn down

knopp replied:
I meant the overall flow. For example destroying the window from Flutter involoves sending `WM_CLOSE`. But WM_CLOSE is the one message we should actually handle, prevent default and notify the framework so that we can give the user a way to prevent closing of windows if needed.

hbatagelo replied:
`WM_CLOSE` is handled by the app lifecycle manager, but it only works for the last window. We expected to have a proper window lifecycle manager for this. Anyway, the delegate pattern you're using in the FFI branch seems like a nice, cleaner alternative.

knopp commented:
Here is my work in progress branch based on this work but converting it to ffi: https://github.com/knopp/flutter/tree/windows_multi_window_ffi

It's bit rough but it mostly works.

If you compare the [`window.dart`](https://github.com/knopp/flutter/blob/windows_multi_window_ffi/packages/flutter/lib/src/widgets/window.dart) to original you will notice that it doesn't need any futures, `isReady` checks or future builders.

It also doesn't need to need to sync state from native (i.e. size). There is only one source of truth and that is whatever winapi provides.

The prototype also allows handling win32 messages in [dart code](https://github.com/knopp/flutter/blob/915c7bb2dbc771159fd19bda78acade8f50a55a9/packages/flutter/lib/src/widgets/window_win32.dart#L183-L195). Down the line this would allow for packages to customized window behavior, i.e. overriding `WM_NCHITTEST` which is required to make custom frame windows.

hbatagelo committed:
Merge branch 'master' into foundation

hbatagelo committed:
Simplify focusing the root view when the top-level window is activated

mattkae committed:
Simplify focusing the root view when the top-level window is activated (#37)

This simplifies the logic for focusing the root view when the top-level
window is activated. It removes the handling of `WM_MOUSEACTIVATE`, as
`WM_ACTIVATE` already accounts for activations via mouse clicks.
Additionally, it ensures that the view is not focused when the top-level
window is being deactivated.

mattkae committed:
Merge branch 'master' into foundation

mattkae committed:
bugfix: get rid of lifecycle state

mattkae committed:
Merge branch 'foundation' of github.com:canonical/flutter into foundation

hbatagelo committed:
Remove unused key and add offset type for relative_position

knopp commented:
I think down the line we want to refer to size as clientSize or contentSize (I'd vote for contentSize). Some platforms support setting frame size too, and we will want to expose it, even if it's per platform specific call. So it would not good to get rid of ambiguity from beginning.

hbatagelo committed:
Merge branch 'master' into foundation

hbatagelo committed:
Merge branch 'master' into foundation

wanjm commented:
how much work left before it can be merged?

Saviq commented:
@wanjm there's a new approach to crossing the Dart / native boundary in the works:

https://github.com/flutter/flutter/pull/165072
https://docs.google.com/document/d/1UEaxd_aIlsPMzkfB6oyeYaKzwHadrWGyMzOS33oSbK4/edit

Not too long! We have agreement on the Dart API, multi-view prerequisites mostly dealt with, just need to work things out on the three desktop platforms so we're confident with the solution.

GroovinChip commented:
> @wanjm there's a new approach to crossing the Dart / native boundary in the works:
> 
> #165072 https://docs.google.com/document/d/1UEaxd_aIlsPMzkfB6oyeYaKzwHadrWGyMzOS33oSbK4/edit
> 
> Not too long! We have agreement on the Dart API, multi-view prerequisites mostly dealt with, just need to work things out on the three desktop platforms so we're confident with the solution.

Is the plan to move over to this approach, or to continue on with method channels until the FFI approach is finalized and ready? If the plan is to switch to FFI now, does that impact the time until multi-window can be delivered?

Saviq commented:
> Is the plan to move over to this approach, or to continue on with method channels until the FFI approach is finalized and ready? If the plan is to switch to FFI now, does that impact the time until multi-window can be delivered?

Pan is to go for FFI, no point introducing a lot of code that will be dropped soon. Yes, it does mean it's delayed some weeks. But it's closer than you might think: #165072.

GroovinChip commented:
Thanks for the reply! Glad the delay will only be a few weeks.

mattkae commented:
The implementation of multi-window in Flutter has drifted a lot since this PR was opened. Please see https://github.com/flutter/flutter/pull/168728 for the latest implementation. We're currently focusing on an FFI-based approach in contrast to the async method channels approach that we implemented here.
